README.md
[![crates.io](https://img.shields.io/crates/v/arloader.svg)](https://crates.io/crates/arloader)
[![build status](https://github.com/CalebEverett/arloader/actions/workflows/build.yml/badge.svg)](https://github.com/CalebEverett/arloader/actions/workflows/build.yml)
[![docs.rs](https://img.shields.io/docsrs/arloader)](https://docs.rs/arloader)
![Crates.io](https://img.shields.io/crates/l/arloader)

# arloader

Command line application and client for uploading files to [Arweave](https://www.arweave.org/). Arweave stores documents and applications forever.

Upload gigabytes of files with one command. Files are read and posted to [arweave.net](https://arweave.net) asynchronously with computationally intensive bundle preparation performed in parallel on multiple threads.

## Contents
* [Installation](#installation)
* [NFT Usage](#nft-usage)
* [General Usage](#general-usage)
* [Usage with SOL](#usage-with-sol)
* [Reward Multiplier](#reward-multiplier)
* [Usage without Bundles](#usage-without-bundles)
* [Benchmarks](#benchmarks)
* [Pricing Comparison](#pricing-comparison)
* [Roadmap](#roadmap)
* [Transactions Prior to v0.1.51](#potential-issue-with-transactions-uploaded-prior-to-version-0.1.51)

## Installation

1. The easiest way to use arloader is to download the binary for your system (Linux or Mac) from the [releases on github](https://github.com/CalebEverett/arloader/releases).

You can also install from [crates.io](https://crates.io) once you have [rust installed](https://www.rust-lang.org/tools/install) with the nightly toolchain.

```
rustup default nightly
cargo install arloader
```

2. Get an Arweave wallet json file [here](https://faucet.arweave.net/).

3. If you're going to use AR to pay for transactions, [get AR tokens](https://arweave.news/how-to-buy-arweave-token/).

4. If you're going to use SOL, get a [Solana wallet](https://docs.solana.com/wallet-guide/cli) json file and transfer some SOL to it.

## NFT Usage
The single `upload-nfts` command below assumes you have a pair of image and metadata files for each of your NFTs. To learn more about each of the steps to upload your files, check out [upload_nfts_steps](docs/upload_nfts_steps.md). See [multiple_asset_files](docs/multiple_asset_files.md) for an example of how to upload multiple media files.

For a beginner's step-by-step guide of how to create an NFT in Solana (from creating a wallet, to uploading assets to arweave using arloader, to listing it in a marketplace) check out [this medium article](https://medium.com/@murki/the-ultimate-dev-guide-to-manually-minting-a-brand-new-nft-in-solana-fb5af9771688).

### Create Upload Folder
 Put your assets and associated metadata files with `.json` extension in a folder by themselves. You can use any kind of file you want. Arloader automatically adds a content type tag to your upload so that browsers will handle it correctly when accessed from Arweave.
 ```
├── 0.json
├── 0.png
├── 1.json
├── 1.png
├── 2.json
├── 2.png
├── 3.json
├── 3.png
├── 4.json
├── 4.png
├── 5.json
├── 5.png
```

### Upload Assets
If you want to fund transactions with SOL, run the command below where `<FILE_PATHS>` matches your asset files.
```
arloader upload-nfts <FILE_PATHS> --with-sol --sol-keypair-path <SOL_KEYPAIR_PATH> --ar-default-keypair
```

For example, if you were uploading mp4 files as your assets and they were in a folder called `path/to/my/assets` and the path to your SOL keypair was `path/to/my/solkeypair.json`, you would enter:

```
arloader upload-nfts path/to/my/assets/*.mp4 --with-sol --sol-keypair-path path/to/my/solkeypair.json --ar-default-keypair
```

To fund transactions with AR, instead run:
```
arloader upload-nfts <FILE_PATHS> --ar-keypair-path <AR_KEYPAIR_PATH>
```

This will first upload your assets, logging statuses to a newly created directory named `arloader_<RANDOM_CHARS>` in the folder where the assets are located.

Then a manifest file will be created from the logged statuses and uploaded. A manifest is a special file that Arweave uses to access your files by their names, relative to the id of the manifest transaction: `https://arweave.net/<MANIFEST_ID>/<FILE_PATH>`. You'll still be able to access your files by their id at `https://arweave.net/<BUNDLE_ITEM_ID>`, but creating and uploading a manifest gives you the option of using either. Once uploaded, the manifest file itself can be accessed online at `https://arweave.net/tx/<MANIFEST_ID>/data.json`.

#### Update Metadata and Upload 
Next your metadata files will be updated with links to the uploaded assets. For the `upload-nfts` command arloader will replace the `image` key with the newyly created link and append the new link(s) to `files` key in your metadata `.json`. It defaults to using the id link, `https://arweave.net/<BUNDLE_ITEM_ID>`, but if you prefer to use the file path based link, `https://arweave.net/<MANIFEST_ID>/<FILE_PATH>`, you can pass the `--link-file` flag.

After your metadata files have been updated, they will be uploaded, followed by the creation and upload of a manifest file for your metadata  files.

### Get Links to Uploaded Metadata

Once everything has been uploaded, the links to your uploaded metadata files, to be included in your on chain token metadata, can be found in `arloader_<RAND_CHAR>/metadata/manifest_<TXID>.json`.

```json
{
    "0.json": {
        "id": "ScU9mEuKBbPX5o5nv8DZkDnZuJbzf84lyLk-uLVDqNk",
        "files": [
            {
                "uri": "https://arweave.net/ScU9mEuKBbPX5o5nv8DZkDnZuJbzf84lyLk-uLVDqNk",
                "type": "application/json"
            },
            {
                "uri": "https://arweave.net/fo9P3OOq78REajk48vFWbKfIhw6mDzgjANQIh3L7Njs/0.json",
                "type": "application/json"
            }
        ]
    },
    "1.json": {
        "id": "8APeQ5lW0-csTcBaGdPBDLAL2ci2AT9pTn2tppGPU_8",
        "files": [
            {
                "uri": "https://arweave.net/8APeQ5lW0-csTcBaGdPBDLAL2ci2AT9pTn2tppGPU_8",
                "type": "application/json"
            },
            {
                "uri": "https://arweave.net/fo9P3OOq78REajk48vFWbKfIhw6mDzgjANQIh3L7Njs/1.json",
                "type": "application/json"
            }
        ]
    },
```

If you are creating your NFTs with the [Metaplex Candy Machine](https://docs.metaplex.com/candy-machine-v2/introduction), you can use the file named `metaplex_items_<MANIFIEST_ID>.json` in the same directory as `<MANIFEST_PATH>`. As with updating metadata files, Arloader defaults to using the id based link, `https://arweave.net/<BUNDLE_ITEM_ID>`, but you can use the file based link, `https://arweave.net/<MANIFEST_ID>/<FILE_PATH>`, by passing the `--link-file` flag.

```json
{
        "0": {
            "link": "uri link",
            "name": "name",
            "onChain": false
        },
        "1": {
            "link": "uri link",
            "name": "name",
            "onChain": false
        },
```
### Confirm All Transactions

Before you create your tokens, make sure that all of your transactions have been confirmed at least 25 times. Run the command below where `<LOG_DIR>` refers to the automatically created directory in your assets folder that begins with `arloader_`.

The primary reason for transactions having a status of `NotFound` is that they were rejected by miners for not having a big enough reward or for having to big of a reward. See [Reward Multiplier](#reward-multiplier) for instructions on increasing the reward by passing the optional `--reward-multiplier` argument.

```
arloader update-nft-status <LOG_DIR>
```

```
Updating asset bundle statuses...

 bundle txid                                   items      KB  status       confirms
------------------------------------------------------------------------------------
 kmgLCgV-dB-DGML8cvFwuP3a-ZKedz7nyDuEsqYPTis      10     980  Confirmed          60
Updated 1 statuses.


Updating metadata bundle statuses...

 bundle txid                                   items      KB  status       confirms
------------------------------------------------------------------------------------
 kmgLCgV-dB-DGML8cvFwuP3a-ZKedz7nyDuEsqYPTis      10     980  Confirmed          60
Updated 1 statuses.


Updating asset manifest status...

 id                                           status     confirms
------------------------------------------------------------------
 URwQtoqrbYlc5183STNy3ZPwSCRY4o8goaF7MJay3xY  Confirmed        60


Updating metadata manifest status...

 id                                           status     confirms
------------------------------------------------------------------
 fo9P3OOq78REajk48vFWbKfIhw6mDzgjANQIh3L7Njs  Confirmed        57
 ```

## General Usage

If you're uploading more than one file, you should pretty much always be using bundles. Bundles take multiple files and packages them together in a single transaction. This is better than uploading multiple individual files because you only have to wait for one transaction to be confirmed. Once the bundle transaction is confirmed, all of your files will be available. Larger transactions with larger rewards are more attractive to miners, which means a larger bundled transaction is more likely to get written quickly than a bunch of smaller individual ones.

Arloader will create as many bundles as necessary to upload all of your files. Your files are read asynchronously, bundled in parallel across multiple threads and then posted to [arweave.net](https://arweave.net). Arloader supports bundle sizes up to 200 MB, with a default of 100 MB. This should work fine for individual files up to 100 MB. If your files sizes are bigger than 100 MB (but smaller than 200 MB), you can specify a larger bundle size with the `--bundles-size` argument - `--bundle-size 200` to specify a size of 200 MB, for example. If your file sizes are bigger than 200 MB, you can upload them as individual files by passing the `--no-bundle` flag.

### Estimate Cost
To get an estimate of the cost of uploading your files run

```
arloader estimate <FILE_PATHS>
```

`<FILE_PATHS>` can be a glob, like `path/to/my/files/*.png`, or one or more files separated by spacees, like `path/to/my/files/2.mp4 path/to/my/files/0.mp path/to/my/files/2.mp`.

### Upload
To upload your files run

```
arloader upload <FILE_PATHS>
```

This kicks off the process of uploading a stream of bundles created from your files. The example output below had a bundle size of 5000 bytes.

```
bundle txid                                   items      KB  status       confirms
------------------------------------------------------------------------------------
 QGPFcZq91lQgmmz2l7rQHkSQpgfJi-Vhv47oTqIYLm4       2       3  Submitted           0
 _-bhdsi4irDEWz8R9wXT-1c06WVQVSMAmQxVF9OkW94       2       3  Submitted           0
 -OAWdFiGS4NKOZXVJG3yZ0yN4xydGOhfQGX2FCdlG88       2       3  Submitted           0
 UBWGFKyTrUVaCa7wi_181FjAd545vdoHmBQEdlaVdA4       2       3  Submitted           0
 qzQlASZrQXNF9HYIOTPjEZL9uy1U9Ou086kCkQWqld0       2       3  Submitted           0
 ```

A status object gets written to a json file named `<TXID>.json` in a newly created sub directory in the parent folder of the first file in `<FILE_PATHS>`. The folder will be named `arloader_<RAND_CHAR>`. You can specify an existing folder to write statuses to by passing the `--log-dir` argument.

```json
{
    "id": "_-bhdsi4irDEWz8R9wXT-1c06WVQVSMAmQxVF9OkW94",
    "status": "Submitted",
    "file_paths": {
        "tests/fixtures/8.png": {
            "id": "0jd-NTQUZhmnKRY-kMt2vEcmSqgzKOLX_P3QYw6CaNE"
        },
        "tests/fixtures/9.png": {
            "id": "1XdiLkoZ5POHsNx7eLyRgisjnxTLzW8SxGsRcb22j84"
        }
    },
    "number_of_files": 2,
    "data_size": 3546,
    "created_at": "2021-11-23T05:47:41.948103600Z",
    "last_modified": "2021-11-23T05:47:41.948107100Z",
    "reward": 50947968
}
```

### Check Status
After uploading your files, you'll want to check on their status to make sure the have been uploaded successfully and that they ultimately are confirmed at least 25 times before you can be absolutely certain they have been permanently uploaded.

```
arloader update-status <LOG_DIR>
```

This will read the files in `<LOG_DIR>`, looking for a valid transaction id as a file stem, and then go out to the network to update the status of each. The example below contained two sets of bundles, one still pending and one with 45 confirmations.

```
bundle txid                                   items      KB  status       confirms
------------------------------------------------------------------------------------
 -OAWdFiGS4NKOZXVJG3yZ0yN4xydGOhfQGX2FCdlG88       2       3  Pending             0
 _-bhdsi4irDEWz8R9wXT-1c06WVQVSMAmQxVF9OkW94       2       3  Pending             0
 qzQlASZrQXNF9HYIOTPjEZL9uy1U9Ou086kCkQWqld0       2       3  Pending             0
 QGPFcZq91lQgmmz2l7rQHkSQpgfJi-Vhv47oTqIYLm4       2       3  Pending             0
 UBWGFKyTrUVaCa7wi_181FjAd545vdoHmBQEdlaVdA4       2       3  Pending             0
 KuuEZpbfCbw6izMeN3knWlpzmaFhnrDL9dUKCW2LQHw       2       3  Confirmed          45
 IRToYYvsftCiR71sfW5qt8XCzBFotwoDFBoEMEtrMrU       2       3  Confirmed          45
 M2QZYxUqw3ZJ2KXzU4pfw9fFIkVOSrJbSpE7NAvHLvo       2       3  Confirmed          45
 qvci4i6Mfr-5_NHI1bL-Omv16QEUw3iiirzv4fXefnM       2       3  Confirmed          45
 NAP2vTKQdMG_eKyKBYz3876T4yBFl4oYFYqwwwnHbFA       2       3  Confirmed          45
 ```

### Re-Upload
If you find that not all of your transactions have a status of `Confirmed` or that the number of confirmations is below 25 after some period of time, you will want to re-upload your transactions with the following command:

```
arloader reupload <FILE_PATHS> --log-dir <LOG_DIR> --statuses <STATUSES> --max-confirmations <MAX_CONFIRMATIONS>
```
This will first check to make sure that all of the files in `<FILE_PATHS>` are included in the status objects in `<LOG_DIR>`, adding them to the list of files to be reuploaded. Then it will filter the paths in the status objects based on `<STATUSES>` and `<MAX_CONFIRM>`. You can provide multiple statuses and max confirmations will only re-upload transactions with fewer than `<MAX_CONFIRM>` confirmations.

For example, if had uploaded a bunch of jpegs that were in `my/images` and statuses had been logged to `my/images/arloader_hehQJu-RJpo`, if you wanted to re-upload transactions with either a status of `NotFound` or `Pending`, you would run:

```
arloader reupload my/images/*.jpeg --log-dir my/images/arloader_hehQJu-RJpo --statuses NotFound Pending
```

If you wanted to reupload anything with less than 25 confirmations, you would run:

```
arloader reupload my/images/*.jpeg --log-dir my/images/arloader_hehQJu-RJpo --max-confirms 25
```


### Create Manifest

CHANGELOG.md
# Changelog

All notable changes starting with v0.1.34 to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

# v0.1.63 (2022-05-11)
- **changed:** removed default enum feature to make it possible to use stable tool chain.

# v0.1.61 (2022-03-13)
- **changed:** removed trailing slash from `tx/` post to be compatible with change in arweave gateway.

# v0.1.60 (2022-03-08)
- **changed:** `get_status` no longer panics on unspecified arweave response status, but instead returns and error containing the unspecified status.

# v0.1.59 (2022-02-26)
- **added:** add logic to retry uploading chunks until they receive a status 200 response.
- **changed:** increase upload buffer size to 5 now that chunks are retried if not 200 response.

# v0.1.58 (2022-02-24)
- **changed:** reduced default upload buffer size to 1 to try to eliminate bundle verification errors.

# v0.1.57 (2022-02-15)
- **changed:** reduced max `--reward-multiplier` from 10.0 to 3.0 since Arweave transactions seemed to fail for overspend greater than that.
- **added:** implemented `Send` and `Sync` for `Error` and [error_send](examples/error_send.rs) example.

# v0.1.56 (2022-02-07)
- **changed:** increase default bundle size to 100 MB to reduce number of tx and use `chunk/` endpoint instead of `tx/` endpoint used for bundles 10 MB or less.
- **changed:** post transactions in chunks when uploading files greater than 10 MB to allow individual files of unlimited size to be uploaded.
- **added:** include [upload_files](examples/upload_files.rs) example.

# v0.1.55 (2022-01-29)
- **added:** include `write-metaplex-items` in `upload-nfts` command.

# v0.1.54 (2022-01-18)
- **added:** added `--update-animation-url` flag to `update-metadata`.
- **added:** added [instructions](docs/multiple_asset_files.md) for uploading with multiple asset files.
- **changed:** validate `<FILE_PATHS>` for `write-metaplex-items` command as json and fix docs to specify json instead of assets.

# v0.1.53 (2022-01-12)
- **fixed:** Include files as objects with `uri` and `type` keys in the `files` key of the `properties` key when updating metadata files.
- **changed:** File objects are now appended to the `files` key with the `update-metadata` command and in order to update the `image` key, you have to pass the `--update-image` flag. This was done to make it possible to have multiple asset files and associated links for each NFT, in which case you can run `update-metadata` multiple times without the `--update-image` flag to append multiple links to the `files` key.

# v0.1.52 (2022-01-04)
- **changed:** Update fees to floor of 10,000 lamports and rate of 2500 winstons per lamport.
- **changed:** Fixed write-metaplex-items command.

# v0.1.51 (2021-12-23)
- **changed:** change chunking for calculating data roots.

# v0.1.50 (2021-12-17)
- **added:** pricing comparison example.

# v0.1.49 (2021-12-16)
- **changed:** removed file based link from metadata files property when not passing `--link-file` to `update-metadata`.

# v0.1.48 (2021-12-15)
- **added:** added examples to cli `--help`.
- **removed:** remove Windows build.

# v0.1.46 (2021-12-15)
- **added:** added `--reupload` command for bundles. You specify `<FILE_PATHS>` and `--log-dir` and the command reuploads any files not included in the statuses in `--log-dir` and any files with a status included in `--statuses` or fewer confirmations than `--max-confirms`.
- **changed:** changed the way file paths are specified from glob strings to file paths. the `--glob` argument has been replaced by `--file-paths`. This new approach has a better user interface and is more robust across operating systems. Instead of avoiding shell expansion by wrapping glob arguments in quotes, the new approach takes advantage of it. It is now possible to specify an individual file path, a list of file paths separated by spaces or a glob pattern that gets expanded by the shell into multiple space separated file paths.
- **changed:** renamed `--upload-filter` subcommand to `--re-upload`.
- **changed:** refactored `--update-status` to include a long name for `<GLOB>`, `--glob`, and to make it required with `--no-bundle`.

# v0.1.45 (2021-12-13)
- **changed:** cleaned up cli text.
- **added:** added message to handle error when Solana network is unavailable.

# v0.1.44 (2021-12-13)
- **fix:** duration on retry sol service too long.

# v0.1.43 (2021-12-12)
- **fix:** bug adding trailing slash to glob strings in addition to directory strings.
- **changed:** made `--log-dir` an optional argument for `upload`, defaulting instead to creating a sub directory in the parent dir of the first file matching `glob` named `arloader_<RAND_CHARS>`.
- **changed:** removed `--log-dir` argument from `write-metaplex-items` in favor of writing to same parent director as `--manifest-path`.
- **changed:** removed requirement to pass `--log-dir` ahead of value in commands where it is the first argument, including `upload-manifest` and `update-status`.

# v0.1.42 (2021-12-11)
- **fixed:** `upload_nfts` example to create `target/examples/upload_nfts` 

# v0.1.41 (2021-12-11)
- **added:** `--upload-nfts` command that automates uploading of pairs of assets and metadata files, including updating metadata files with links to uploaded assets. You can now provide a glob pattern matching your asset files and the complete process will run, returning a manifest file with links to your uploaded metadata files that can be included in your on chain token metadata.
- **added:** `--update-nft-status` command that reports on status for all nft uploads, assets, metadata and manifests.
for both assets and metadata.
- **added:** `examples/upload_nfts`.
- **changed:** manifest paths now just include just file name instead of the full relative path.
- **changed:** sol service now has private rpc.


src/bundles.rs
use crate::error::Error;
use crate::transaction::{Base64, Tag};
use avro_rs::Schema;
use bytes::BufMut;
use serde::{Deserialize, Serialize};
use std::io::Write;

pub fn get_tags_schema() -> Schema {
    let schema = r#"
        {
            "type": "array",
            "items": {
                "type": "record",
                "name": "tag",
                "fields": [
                    {"name": "name", "type": "string"},
                    {"name": "value", "type": "string"}
                ]
            }
        }
    "#;

    let schema = Schema::parse_str(schema).unwrap();

    schema
}

#[derive(Debug, Serialize, Deserialize, PartialEq)]
pub struct DataItem {
    signature_type: u16,
    signature: Base64,
    owner: Base64,
    target: Option<Base64>,
    anchor: Option<Base64>,
    tags: Vec<Tag>,
    data: Base64,
}

impl Default for DataItem {
    fn default() -> Self {
        Self {
            signature_type: 1,
            signature: Base64(vec![0; 512]),
            owner: Base64(vec![0; 512]),
            target: None,
            anchor: None,
            tags: Vec::new(),
            data: Base64::default(),
        }
    }
}

impl DataItem {
    pub fn serialize(&self) -> Result<Vec<u8>, Error> {
        let mut buf = Vec::new().writer();
        buf.write(&self.signature_type.to_le_bytes())?;
        buf.write(&self.signature.0)?;
        buf.write(&self.owner.0)?;
        if let Some(target) = &self.target {
            buf.write(&[1])?;
            buf.write(&target.0)?;
        } else {
            buf.write(&[0])?;
        }
        if let Some(anchor) = &self.anchor {
            buf.write(&[1])?;
            buf.write(&anchor.0)?;
        } else {
            buf.write(&[0])?;
        }

        if self.tags.len() > 0 {
            let number_of_tags = self.tags.len() as u64;
            let schema = get_tags_schema();
            let value = avro_rs::to_value(&self.tags)?;
            let tags_bytes = avro_rs::to_avro_datum(&schema, value)?;
            let number_of_tag_bytes = tags_bytes.len() as u64;

            buf.write(&number_of_tags.to_le_bytes())?;
            buf.write(&number_of_tag_bytes.to_le_bytes())?;
            buf.write(&tags_bytes)?;
        } else {
            buf.write(&[0; 16])?;
        };

        buf.write(&self.data.0)?;

        Ok(buf.into_inner())
    }
    pub fn deserialize(bytes_vec: Vec<u8>) -> Result<Self, Error> {
        let mut iter = bytes_vec.into_iter();
        let mut data_item = DataItem::default();

        let result = [(); 2].map(|_| iter.next().unwrap());
        data_item.signature_type = u16::from_le_bytes(result);

        data_item.signature.0 = data_item
            .signature
            .0
            .iter()
            .map(|_| iter.next().unwrap())
            .collect();

        data_item.owner.0 = data_item
            .owner
            .0
            .iter()
            .map(|_| iter.next().unwrap())
            .collect();

        data_item.target = match iter.next().unwrap() {
            0 => None,
            1 => Some(Base64([(); 32].map(|_| iter.next().unwrap()).to_vec())),
            _ => unreachable!(),
        };

        data_item.anchor = match iter.next().unwrap() {
            0 => None,
            1 => Some(Base64([(); 32].map(|_| iter.next().unwrap()).to_vec())),
            _ => unreachable!(),
        };

        let result = [(); 8].map(|_| iter.next().unwrap());
        let number_of_tags = u64::from_le_bytes(result);
        let result = [(); 8].map(|_| iter.next().unwrap());
        let number_of_tag_bytes = u64::from_le_bytes(result) as usize;
        println!("{:?}", &number_of_tag_bytes);

        data_item.tags = if number_of_tags > 0 {
            let schema = get_tags_schema();
            let mut reader = Vec::<u8>::with_capacity(number_of_tag_bytes);

            for _ in 0..number_of_tag_bytes {
                reader.push(iter.next().unwrap());
            }

            let value = avro_rs::from_avro_datum::<&[u8]>(&schema, &mut &*reader, None)?;
            let tags: Vec<Tag> = avro_rs::from_value(&value)?;
            tags
        } else {
            Vec::<Tag>::new()
        };

        data_item.data.0 = iter.collect();

        Ok(data_item)
    }
}

#[cfg(test)]
mod tests {
    use super::DataItem;
    use crate::{
        crypto::Provider,
        transaction::{Base64, ConvertUtf8, FromStrs, Tag},
    };
    use std::path::PathBuf;

    async fn get_test_data_item() -> DataItem {
        let crypto = Provider::from_keypair_path(PathBuf::from("tests/fixtures/test_key0.json"))
            .await
            .unwrap();

        let tags = vec![
            Tag::from_utf8_strs(&"e".to_string().repeat(56), "testvalue").unwrap(),
            Tag::from_utf8_strs(&"e".to_string().repeat(56), "testvalue").unwrap(),
        ];
        let owner = crypto.keypair_modulus().unwrap();
        let anchor = Some(Base64::from_utf8_str("TWF0aC5hcHQnI11nbmcoMzYpLnN1YnN0").unwrap());
        let data = Base64::from_utf8_str("tasty").unwrap();

        DataItem {
            owner,
            anchor,
            tags,
            data,
            ..Default::default()
        }
    }

    #[tokio::test]
    async fn test_serialize_data_item() {
        let data_item = get_test_data_item().await;

        let bytes = data_item.serialize().unwrap();

        let expected_bytes: Vec<u8> = vec![
            1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
            0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
            0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
            0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
            0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
            0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
            0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
            0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
            0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
            0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
            0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
            0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
            0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
            0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
            0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
            0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
            0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
            0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 232, 198, 246, 244, 49,
            105, 123, 28, 223, 239, 90, 240, 124, 139, 75, 157, 113, 26, 141, 150, 198, 202, 113,
            248, 196, 220, 207, 158, 154, 5, 189, 26, 25, 188, 59, 67, 105, 63, 55, 133, 189, 123,
            158, 253, 50, 255, 247, 134, 22, 115, 101, 211, 178, 254, 131, 106, 214, 43, 176, 242,
            168, 154, 9, 107, 44, 52, 254, 216, 15, 52, 12, 23, 174, 195, 236, 59, 52, 103, 241,
            167, 70, 78, 52, 247, 48, 11, 243, 99, 198, 75, 16, 65, 101, 48, 124, 98, 226, 12, 195,
            52, 160, 250, 166, 208, 211, 136, 184, 13, 18, 168, 178, 224, 75, 223, 92, 227, 55,
            181, 109, 144, 119, 131, 108, 106, 0, 23, 0, 74, 113, 247, 221, 14, 98, 186, 113, 176,
            108, 197, 79, 133, 254, 59, 151, 243, 39, 124, 225, 191, 171, 174, 40, 200, 255, 70,
            90, 172, 223, 66, 56, 124, 202, 198, 176, 125, 198, 84, 139, 180, 89, 48, 207, 239,
            131, 66, 110, 86, 120, 114, 182, 208, 95, 174, 185, 253, 112, 255, 105, 176, 95, 219,
            129, 55, 218, 84, 200, 153, 196, 145, 184, 180, 95, 38, 86, 91, 54, 122, 127, 131, 218,
            12, 205, 213, 111, 1, 136, 253, 236, 123, 232, 54, 109, 251, 33, 82, 59, 97, 195, 6,
            111, 183, 225, 54, 244, 53, 130, 151, 10, 39, 40, 197, 43, 39, 133, 168, 213, 254, 68,
            150, 67, 95, 39, 131, 249, 201, 208, 82, 151, 48, 146, 24, 139, 161, 157, 90, 164, 64,
            21, 151, 138, 166, 34, 109, 215, 73, 229, 181, 89, 134, 233, 104, 89, 238, 251, 141,
            19, 58, 112, 116, 190, 125, 23, 171, 134, 242, 80, 104, 190, 132, 150, 48, 204, 71,
            155, 94, 195, 52, 158, 84, 72, 137, 156, 214, 202, 52, 208, 201, 146, 104, 138, 250,
            28, 187, 226, 93, 191, 246, 71, 116, 230, 25, 233, 190, 18, 212, 78, 168, 59, 153, 209,
            226, 144, 197, 249, 89, 95, 100, 195, 11, 90, 23, 114, 104, 194, 102, 45, 246, 105, 90,
            192, 122, 2, 92, 206, 41, 30, 244, 250, 131, 195, 105, 233, 92, 200, 2, 164, 27, 208,
            239, 73, 143, 187, 58, 100, 170, 53, 96, 2, 250, 120, 135, 198, 196, 229, 136, 194, 36,
            235, 10, 208, 26, 111, 145, 189, 234, 124, 156, 37, 201, 10, 29, 123, 162, 231, 112,
            172, 82, 249, 65, 253, 237, 24, 99, 178, 52, 45, 208, 157, 212, 126, 221, 29, 148, 5,
            188, 121, 51, 67, 34, 99, 22, 104, 94, 157, 142, 115, 53, 157, 12, 147, 133, 24, 41,
            202, 141, 232, 19, 223, 113, 25, 22, 121, 144, 152, 9, 221, 160, 18, 254, 27, 186, 120,
            250, 173, 187, 132, 13, 203, 206, 176, 86, 83, 119, 176, 182, 36, 91, 65, 243, 193,
            241, 124, 79, 123, 71, 38, 238, 172, 58, 152, 47, 130, 205, 106, 101, 36, 43, 5, 155,
            159, 170, 247, 37, 0, 1, 84, 87, 70, 48, 97, 67, 53, 104, 99, 72, 81, 110, 73, 49, 49,
            110, 98, 109, 99, 111, 77, 122, 89, 112, 76, 110, 78, 49, 89, 110, 78, 48, 2, 0, 0, 0,
            0, 0, 0, 0, 182, 0, 0, 0, 0, 0, 0, 0, 4, 150, 1, 90, 87, 86, 108, 90, 87, 86, 108, 90,
            87, 86, 108, 90, 87, 86, 108, 90, 87, 86, 108, 90, 87, 86, 108, 90, 87, 86, 108, 90,
            87, 86, 108, 90, 87, 86, 108, 90, 87, 86, 108, 90, 87, 86, 108, 90, 87, 86, 108, 90,
            87, 86, 108, 90, 87, 86, 108, 90, 87, 86, 108, 90, 87, 86, 108, 90, 87, 86, 108, 90,
            87, 86, 108, 90, 87, 85, 24, 100, 71, 86, 122, 100, 72, 90, 104, 98, 72, 86, 108, 150,
            1, 90, 87, 86, 108, 90, 87, 86, 108, 90, 87, 86, 108, 90, 87, 86, 108, 90, 87, 86, 108,
            90, 87, 86, 108, 90, 87, 86, 108, 90, 87, 86, 108, 90, 87, 86, 108, 90, 87, 86, 108,
            90, 87, 86, 108, 90, 87, 86, 108, 90, 87, 86, 108, 90, 87, 86, 108, 90, 87, 86, 108,
            90, 87, 86, 108, 90, 87, 86, 108, 90, 87, 86, 108, 90, 87, 85, 24, 100, 71, 86, 122,
            100, 72, 90, 104, 98, 72, 86, 108, 0, 116, 97, 115, 116, 121,
        ];

        assert_eq!(&bytes, &expected_bytes);
    }

    #[tokio::test]
    async fn test_deserialize_data_item() {
        let data_item = get_test_data_item().await;

        let bytes = data_item.serialize().unwrap();

        let de_data_item = DataItem::deserialize(bytes).unwrap();

        assert_eq!(data_item, de_data_item)
    }

    #[tokio::test]
    async fn test_data_item_to_json() {
        let data_item = get_test_data_item().await;
        println!("{}", serde_json::to_string(&data_item).unwrap());
        assert_eq!(
            data_item.data.to_utf8_string().unwrap(),
            "tasty".to_string()
        );
    }
}
